src/controllers/howDoYouWantToPay.controller.ts

```diff
@@ -0,0 +1,145 @@
+import TYPES from 'app/types'
+import FormValidator from 'app/utils/formValidator.util'
+
+interface ViewModel {
+  data?: Optional<HowDoYouWantToPayModel>
+  errors?: Optional<ValidationErrors>
+}
+
+@controller(HOW_DO_YOU_WANT_TO_PAY_URI)
+export class HowDoYouWantToPayController extends BaseController {
+
+  public constructor(
+    @inject(SessionService) private session: SessionService,
+    @inject(FormValidator) private validator: FormValidator,
+    @inject(DissolutionService) private service: DissolutionService,
+    @inject(PaymentService) private paymentService: PaymentService,
+    @inject(TYPES.PAY_BY_ACCOUNT_FEATURE_ENABLED) private PAY_BY_ACCOUNT_FEATURE_ENABLED: number
+  ) {
+    super()
+  }
+
+  @httpGet('')
+  public async get(): Promise<string | RedirectResult> {
+    const token: string = this.session.getAccessToken(this.httpContext.request)
+    const dissolutionSession: DissolutionSession = this.session.getDissolutionSession(this.httpContext.request)!
+
+    if (!this.PAY_BY_ACCOUNT_FEATURE_ENABLED || await this.isAlreadyPaid(dissolutionSession, token)) {
+      return this.redirect(SEARCH_COMPANY_URI)
+    }
+
+    if (dissolutionSession.paymentType) {
+      const body = {
+        paymentType: dissolutionSession.paymentType
+      } as HowDoYouWantToPayModel
+      return this.renderView(body)
+    }
+
+    return super.render('how-do-you-want-to-pay')
+  }
+
+  @httpPost('')
+  public async post(@requestBody() body: HowDoYouWantToPayModel): Promise<string | RedirectResult>  {
+    if (!this.PAY_BY_ACCOUNT_FEATURE_ENABLED) {
+      return this.redirect(SEARCH_COMPANY_URI)
+    }
+
+    const token: string = this.session.getAccessToken(this.httpContext.request)
+    const dissolutionSession: DissolutionSession = this.session.getDissolutionSession(this.httpContext.request)!
+
+    if (await this.isAlreadyPaid(dissolutionSession, token)) {
+      return this.redirect(SEARCH_COMPANY_URI)
+    }
+
+    const paymentType: PaymentType = body.paymentType
+
+    const errors: Optional<ValidationErrors> = this.validator.validate(body, howDoYouWantToPaySchema)
+    if (errors) {
+      return this.renderView(body, errors)
+    }
+
+    return this.redirect(await this.getRedirectURI(paymentType, dissolutionSession))
+  }
+
+  private async renderView(
+    data?: Optional<HowDoYouWantToPayModel>,
+    errors?: Optional<ValidationErrors>): Promise<string> {
+    const viewModel: ViewModel = {
+      data,
+      errors
+    }
+
+    return super.render('how-do-you-want-to-pay', viewModel, errors ? StatusCodes.BAD_REQUEST : StatusCodes.OK)
+  }
+
+  private async getRedirectURI(paymentType: PaymentType, dissolutionSession: DissolutionSession): Promise<string> {
+    const updatedSession = this.updateSessionPaymentType(dissolutionSession, paymentType)
+
+    if (paymentType === PaymentType.ACCOUNT) {
+      return PAY_BY_ACCOUNT_URI
+    }
+
+    return await this.getCreditCardPaymentUrl(updatedSession)
+  }
```

Comment:
this is doing much more that just getting a URI. Either rename the method or move unrelated logic out of this method into another one

Reply:
Sorry, by 'unrelated logic' are you referring to the session update?

Reply:
yup

Reply:
Moved